3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala

```diff
@@ -265,12 +251,82 @@ object EnrichmentManager {
       }
     }
 
-    // Finally anonymize the IP address
+    // To anonymize the IP address
     Option(event.user_ipaddress).map(ip => event.user_ipaddress = registry.getAnonIpEnrichment match {
       case Some(anon) => anon.anonymizeIp(ip)
       case None => ip
     })
 
+    // Parse the useragent using user-agent-utils
+    val client = {
+      registry.getUserAgentUtilsEnrichment match {
+        case Some(uap) => {
+          Option(event.useragent) match {
+            case Some(ua) =>
+              val ca = uap.extractClientAttributes(ua)
+              ca.flatMap(c => {
+                event.br_name = c.browserName
+                event.br_family = c.browserFamily
+                c.browserVersion.map(bv => event.br_version = bv)
+                event.br_type = c.browserType
+                event.br_renderengine = c.browserRenderEngine
+                event.os_name = c.osName
+                event.os_family = c.osFamily
+                event.os_manufacturer = c.osManufacturer
+                event.dvce_type = c.deviceType
+                event.dvce_ismobile = CU.booleanToJByte(c.deviceIsMobile)
+                c.success
+                })
+              ca
+            case None => unitSuccess // No fields updated
+          }
+        }
+        case None => unitSuccess
+      }
+    }
+
+    // Create the ua_parser_context
+    val uaParser = {
+      registry.getUaParserEnrichment match {
+        case Some(uap) => {
+          Option(event.useragent) match {
+            case Some(ua) => uap.extractUserAgent(ua).map(_.some)
+            case None => None.success // No fields updated
+          }
+        }
+        case None => None.success
+      }
+    }
+
+    // Finalize the currency conversion
+    val currency = {
+      registry.getCurrencyConversionEnrichment match {
+        case Some(currency) => {
+          event.base_currency = currency.baseCurrency
+          val trTax      = CU.stringToMaybeDouble("tr_tx", event.tr_tax).toValidationNel
+          val tiPrice    = CU.stringToMaybeDouble("ti_pr", event.ti_price).toValidationNel
+          val trTotal    = CU.stringToMaybeDouble("tr_tt", event.tr_total).toValidationNel
+          val trShipping = CU.stringToMaybeDouble("tr_sh", event.tr_shipping).toValidationNel
+          val convertedCu = ((trTotal |@| trTax |@| trShipping |@| tiPrice) {
+            currency.convertCurrencies(Option(event.tr_currency), _, _, _, Option(event.ti_currency), _, raw.context.timestamp)
+          }).flatMap(x => x)
+
+          convertedCu foreach {
+            arr => {
+              val (total, tax, shipping, price) = arr
+              event.tr_total_base = total.orNull
```

Comment:
```scala
for ((a,b) <- Some((5, 10)) {}
```

works but

```scala
val x: Validation[NonEmptyList[String], (String,String)] = Success("a","b")
for ((a,b) <- x) {}
```

fails with the error

```
could not find implicit value for parameter M: scalaz.Monoid[scalaz.NonEmptyList[String]]
```

for reasons which I do not understand...